WMF Framework/Widget/WidgetController.swift

```diff
@@ -167,3 +176,96 @@ public final class WidgetController: NSObject {
     #endif
     
 }
+
+/// When the old widget data loading model is removed, this should be moved out of this extension into the class itself and refactored (e.g. the properties here don't need to be computed).
+public extension WidgetController {
+
+    // MARK: - Properties
+
+    fileprivate var widgetCacheDirectoryContainerURL: URL {
+        FileManager.default.wmf_containerURL()
+    }
+
+    fileprivate var widgetCacheDataFileURL: URL {
+        return widgetCacheDirectoryContainerURL.appendingPathComponent("Widget Cache").appendingPathExtension("json")
+    }
+
+    // MARK: - Widget Cache
+
+    /// Returns currently cached `WidgetCache` or an empty default
+    func loadCache() -> WidgetCache {
+        if let data = try? Data(contentsOf: widgetCacheDataFileURL), let decodedCache = try? JSONDecoder().decode(WidgetCache.self, from: data) {
+            return decodedCache
+        }
+
+        return WidgetCache(settings: .default, featuredContent: nil)
+    }
+
+    func saveCache(_ widgetCache: WidgetCache) {
+        let encoder = JSONEncoder()
+        guard let encodedCache = try? encoder.encode(widgetCache) else {
+            return
+        }
+
+        try? encodedCache.write(to: widgetCacheDataFileURL)
+    }
+
+    func updateCacheWith(featuredContent: WidgetFeaturedContent) {
+        var updatedCache = loadCache()
+        updatedCache.featuredContent = featuredContent
+        saveCache(updatedCache)
+    }
+
+    func updateCacheWith(settings: WidgetSettings) {
+        var updatedCache = loadCache()
+        updatedCache.settings = settings
+        saveCache(updatedCache)
+    }
+
+    // MARK: - Featured Article Widget
+
+    func fetchFeaturedContent(isSnapshot: Bool = false, completion: @escaping (WidgetContentFetcher.FeaturedContentResult) -> Void) {
+        func performCompletion(result: WidgetContentFetcher.FeaturedContentResult) {
+            DispatchQueue.main.async {
+                completion(result)
+            }
+        }
+
+        let fetcher = WidgetContentFetcher.shared
+        var widgetCache = loadCache()
+
+        guard !isSnapshot else {
+            performCompletion(result: .success(widgetCache.featuredContent ?? WidgetFeaturedContent(featuredArticle: nil)))
+            return
+        }
+        
+        // If cached data is still relevant, use it
+        if let cachedContent = widgetCache.featuredContent, let fetchDate = cachedContent.fetchDate, Calendar.current.isDateInToday(fetchDate), let cachedLanguageCode = cachedContent.featuredArticle?.languageCode, cachedLanguageCode == widgetCache.settings.languageCode {
+            performCompletion(result: .success(cachedContent))
+            return
+        }
+
+        // Fetch fresh feed content from network
+        fetcher.fetchFeaturedContent(forDate: Date(), siteURL: widgetCache.settings.siteURL, languageCode: widgetCache.settings.languageCode) { result in
```

Comment:
Updated. I was neglecting to use the `languageVariantCode` from the `WidgetSettings` at all, thus the Chinese variants weren't displaying.

I've done two things to rectify this:
1. Updated the `WidgetContentFetcher` to use `languageVariantCode` on fetch requests (which gets added to the `URLRequest`, then added to the `Accept-Language` headers for free via using `Session`)
2. Updated the widget to use `displayTitle` instead of `normalizedTitle`. Looking at some `zh.wikipedia.org` Wikifeeds responses, it appears that `displayTitle` provides the expected variant character set for a request.

Testing with Simplified and Traditional Chinese it appears to now be displaying the correct stuff.

Reply:
This could be related to the `displayTitle` change, but I saw some html in the title for English. It might be worth checking the Explore feed card to see what we're using there. Maybe we're just stripping the html out?

![IMG_0549](https://user-images.githubusercontent.com/3620196/128402671-2bf5ce83-76f5-4c8c-a71a-779ba3dd6241.PNG)

Reply:
🥴 I think that explains why I probably naively went with `normalizedTitle` to begin with. 👍🏽 I'll dig more into it and push up a fix.

Reply:
So [it looks like](https://github.com/wikimedia/wikipedia-ios/blob/32ca59d18352c20419605c6b469cc530f304948d/Wikipedia/Code/ArticleCollectionViewCell.swift#L20) the Featured Article card in the Explore feed converts the `displayTitle`, with its embedded HTML, into an attributed string for display on the card. 

Unfortunately, SwiftUI doesn’t play nicely with the existing attributed string API. While we can certainly write our own lil’ parser to convert the HTML to a SwiftUI-style attributed string for display, that’s probably a heavier than would be worth it just for the sake of this label task. That said, I’m happy to look into and start implementing something if you’d like me to.

I’ve pushed an update that just strips the HTML using our existing `wmf_stringByRemovingHTML` method – lemme know if this works for you or if you want me to start investigating SwiftUI-style attributed string solutions.

Reply:
👍  sounds good, I think stripping html is fine in this case.